### ISSUE TITLE
Weird behavior for root components

### ISSUE THREAD

**Issue by metacatdud** [maxence-charriere/go-app#695]
Hey guys, I have  the weirdest situation:

Given:
```go
type Overview struct {
	app.Compo

	Path             string
	Name             string
	CurrentComponent string

	// DI set
	Sidebar *sidebar.Widget
}
```

I mount it to a route

```go
page := &Overview{
  Path:    "/path",
  Name:    "Name",
  Sidebar: nil,
}

app.Route(page.Path, page)
```

On the `mount` of the component all properties are empty 

```go
func (page *Overview) OnMount(ctx app.Context) {
	fmt.Printf("[MOUNT] %+v\n", page)
	page.init(ctx)
}
```

Outputs:

```shell
[MOUNT] &{Compo:{ ... } Path: Name: CurrentComponent: Sidebar:<nil>}
```
`Path` and `Name` are empty and I think there is something I am missing or I don't understand, because it makes no sense :/

Why is weird? It is because the route itself works, when I call `/path` all work, and the `Render`
 the method is called and displayed correctly, but the receiver no longer has the properties set on it.

```go 
func (page *Overview) Render() app.UI {
	return app.Div().Body( 
```

same as `OnMount`, `page` receiver properties are gone

Thanks

**Comment by oderwat** [comment]
`app.Route()` creates a new instance of your component (`r.routes[path] = reflect.TypeOf(c)`) which is being created and mounted then the route gets active. You need to add an `onNav()` handler, so initialize the data for the component when it is navigated onto. I solve this by only having one route with a router component and all of the page and sidebar components are in a map which I swap in when needed (using my [mountpoint](https://github.com/metatexx/go-app-pkgs/tree/master/mountpoint) package.

**Comment by maxence-charriere** [comment]
Using route only associate an component type with a route. 

When there is navigation, a new instance of the component is created.

https://go-app.dev/routing

You may do component initialization in OnMount or OnNav functions. https://go-app.dev/components

Path, Page title and other header relating thing are accessible from the context given with OnMount or OnNav.

See a component like an independent part with its own lifecycle, that you associate with a path.

**Comment by oderwat** [comment]
@maxence-charriere It is kinda counter-intuitive that it works like this when the documentation tells you that components can be initialized using upper case field names. The wording "... with a given component type" should give it away. Still, it says "Routing is about associating a component with an URL path." before that. I think it should be more clear that this is not how it works when they are added to a route. Maybe extending the example with a section on how to initialize it from the path and mentioning that there is no way to initialize the component before, may help people to understand this part of the system.

**Comment by metacatdud** [comment]
Thanks @maxence-charriere , @oderwat for clearing this out. I dag through the code and found the `reflect.TypeOf(c)`.

I have a small-mid project, ideal candiate, and will not give up until I will make it using go-app. I've been waiting for more than an year for an excuse to use it 😁

There are few drawbacks I have noticed. 
One is this one, which as far as I understand it's more of a problem of understanding and grasping the concept, and the other is the inability to use interfaces. 

I am more of a backend guy, so at first I was prompted to wrap components into interfaace and use a good old even-driven design, but because after pre-render thing happen in client, the reference passed it's not handled correctly

```javasacript
Observer received wrong type
expected: component.Component
received: component.Overview
```
In go `Overview` implements `Component`

I will try some other ways to make this work. The closes I got was with a repo patern of sorts, as in to have a config layer for all components then to use those instances as components.

I saw the project with keygpg. Nice aproach, but the home component render hunction has 2k lines of code sooo, idk if that is a good idea, but in a raw form, my idea it's there.

There are basically 2 layers of interaction:
1. Configuration - where you set a component with it's subcomponents
2. Parametrizing - based on actions and state values a component it's displayed.

So what I am looking for is to make first one happen. The second one, when setting params, it will set a component and that component will distribute data down to its children seamlesly.
